dags/sql_examples/sql_check.py

```diff
@@ -0,0 +1,115 @@
+from airflow import DAG
+from airflow.models.baseoperator import chain
+from airflow.operators.dummy_operator import DummyOperator
+from airflow.utils.dates import datetime
+from airflow.operators.sql import (
+    SQLCheckOperator,
+    SQLValueCheckOperator,
+    SQLIntervalCheckOperator,
+    SQLThresholdCheckOperator
+)
+from airflow.utils.task_group import TaskGroup
+
+import pandas as pd
+
+# This table variable is a placeholder, in a live environment, it is better
+# to pull the table info from a Variable in a template
+TABLE = "yellow_tripdata"
+DATES = ["2019-01", "2019-02"]
+TASK_DICT = {}
+
+# By putting conn_id as a default_arg, the arg is passed to every task,
+# reducing boilerplate
+with DAG("sql_data_quality",
+         start_date=datetime(2021, 7, 7),
+         description="A sample Airflow DAG to perform data quality checks using SQL Operators.",
+         schedule_interval=None,
+         default_args={"conn_id": "postgres_default"},
+         template_searchpath="/usr/local/airflow/include/sql/sql_examples/",
+         catchup=False) as dag:
+    """
+    ### SQL Check Operators Data Quality Example
+
+    Before running the DAG, ensure you have an active and reachable SQL database
+    running, with a connection to that database in an Airflow Connection.
+
+    Note: The data files for this example do **not** include an `upload_date`
+    column. This column is needed for the interval check, and is added as a
+    Task in sql_check_redshift_etl.py.
+    """
+
+    begin = DummyOperator(task_id="begin")
+    end = DummyOperator(task_id="end")
+
+    """
+    #### Run Row-Level Quality Checks
+    For each date of data, run checks on 10 rows to ensure basic data quality
+    cases (found in the .sql file) pass.
+    """
+    for date in DATES:
+        with TaskGroup(group_id=f"row_quality_checks_{date}") as quality_check_group:
+            trip_dict = pd.read_csv(
+                f"/usr/local/airflow/include/data/yellow_tripdata_sample_{date}.csv",
+                header=0,
+                usecols=["vendor_id", "pickup_datetime"],
+                parse_dates=["pickup_datetime"],
+                infer_datetime_format=True
+            ).to_dict()
```

Comment:
Reading files at DAG parse time is a huge warning sign, and something that really needs to be avoided if this dag is trying to illustrate best practices.

(Sometimes it's not avoidable right now sadly. I'm not sure if this is one of those cases or not.)

Reply:
Any recs or links to guides on better practices?

Reply:
Since you're just getting the `vendor_id` and `pickup_datetime` values from the CSVs, you could try doing this in SQL or a SQL procedure if that's available. Get a list of 10 `vendor_id` and `pickup_datetime` combinations from the table in a CTE or if you have this info in some dimensional table and run the same SQL check using those 10 entries in a larger, single query. 

You wouldn't need the `TaskGroup` at that point either presumably.

Reply:
You can pass the date in the new SQL as a `param` if you want to still have separate tasks for each date.

Reply:
So, the `TaskGroup` is still important because each row needs to be queried individually, as the operator only returns one row from a query. So even if the rows are all checked in one query and aggregated in such a way that if one row fails, they all fail, then observability is still limited as only one row will show up in the logs as failing. But, I can use subqueries to at least not do the load.